android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java

```diff
@@ -941,39 +944,55 @@ public static String getAccountName(AccountInfo[] accounts, String address) {
 
-    public static void openTransaction(
-            TransactionInfo txInfo, JsonRpcService jsonRpcService, Activity activity) {
-        if (txInfo == null || txInfo.txHash == null || txInfo.txHash.isEmpty()) {
-            return;
-        }
-        assert jsonRpcService != null;
-        jsonRpcService.getChainId(chainId -> {
-            jsonRpcService.getAllNetworks(networks -> {
-                for (EthereumChain network : networks) {
-                    if (!chainId.equals(network.chainId)) {
-                        continue;
-                    }
-                    String blockExplorerUrl = Arrays.toString(network.blockExplorerUrls);
-                    if (blockExplorerUrl.length() > 2) {
-                        blockExplorerUrl =
-                                blockExplorerUrl.substring(1, blockExplorerUrl.length() - 1)
-                                + "/tx/" + txInfo.txHash;
-
-                        TabUtils.openUrlInNewTab(false, blockExplorerUrl);
-                        TabUtils.bringChromeTabbedActivityToTheTop(activity);
-                        break;
+    public static void openTransaction(TransactionInfo txInfo, JsonRpcService jsonRpcService,
+            AppCompatActivity activity, String accountName) {
+        if (txInfo.txStatus == TransactionStatus.UNAPPROVED) {
+            if (activity instanceof ApprovedTxObserver) {
+                showApproveDialog(txInfo, accountName, activity, ((ApprovedTxObserver) activity));
+                return;
+            }
+            throw new RuntimeException("Activity must implement ApprovedTxObserver");
+        } else {
+            if (txInfo == null || txInfo.txHash == null) {
+                return;
+            }
+            assert jsonRpcService != null;
+            jsonRpcService.getChainId(chainId -> {
+                jsonRpcService.getAllNetworks(networks -> {
+                    for (EthereumChain network : networks) {
+                        if (!chainId.equals(network.chainId)) {
+                            continue;
+                        }
+                        String blockExplorerUrl = Arrays.toString(network.blockExplorerUrls);
+                        if (blockExplorerUrl.length() > 2) {
+                            blockExplorerUrl =
+                                    blockExplorerUrl.substring(1, blockExplorerUrl.length() - 1)
+                                    + "/tx/" + txInfo.txHash;
+                            TabUtils.openUrlInNewTab(false, blockExplorerUrl);
+                            TabUtils.bringChromeTabbedActivityToTheTop(activity);
+                            break;
+                        }
                     }
-                }
+                });
             });
-        });
+        }
+    }
+
+    public static void openTransaction(TransactionInfo txInfo, JsonRpcService jsonRpcService,
+            AppCompatActivity activity, AccountInfo[] accountInfos) {
+        String to = txInfo.txData.baseData.to;
```

Comment:
there is not check for `txInfo == null`, but we do it in the deeper call to `openTransaction`

Reply:
[same here](https://github.com/brave/brave-core/pull/11951#discussion_r789899520)